src/pubsub/ua_pubsub_writergroup.c

```diff
@@ -1096,6 +1112,20 @@ UA_WriterGroup_publishCallback(UA_Server *server, UA_WriterGroup *writerGroup) {
         UA_PublishedDataSet *pds =
             UA_PublishedDataSet_findPDSbyId(server, dsw->connectedDataSet);
         if(!pds) {
+            if (UA_NodeId_isNull(&dsw->connectedDataSet)) {
+                UA_StatusCode res1 =
+                        UA_DataSetWriter_generateDataSetMessageHeartbeat(server,
+                                &dsmStore[dsmCount], dsw);
+                if (res1 != UA_STATUSCODE_GOOD) {
+                    UA_LOG_WARNING(&server->config.logger, UA_LOGCATEGORY_SERVER,
+                            "PubSub Publish: Heartbeat DataSetMessage creation failed");
+                    continue;
+                }
+                dsWriterIds[dsmCount] = dsw->config.dataSetWriterId;
+                dsmCount++;
+                continue;
+            }
+
```

Comment:
Is it possible to check whether its a hearbeat at the same place where we check whether it possible to batch the message?

Reply:
You're right. Fixed by ec25477.